tests/integration/CapabilityTest.php

```diff
@@ -0,0 +1,131 @@
+<?php
+
+namespace Automattic\LegacyRedirector\Tests;
+
+use Automattic\LegacyRedirector\Capability;
+
+/**
+ * CapabilityTest class
+ */
+class CapabilityTest extends TestCase {
+
+	/**
+	 * We remove the Capability::MANAGE_REDIRECTS_CAPABILITY  before testing
+	 *
+	 * @beforeClass
+	 */
+	public static function beforeClassSetup() {
+		$capability = new Capability();
+		$capability->unregister();
+	}
+
+	public function test_admin_capabilities_before_register() {
+		$user_id = $this->factory->user->create( array( 'role' => 'administrator' ) );
+		$user    = wp_set_current_user( $user_id );
+
+		$this->assertFalse( boolval( $user->has_cap( Capability::MANAGE_REDIRECTS_CAPABILITY ) ) );
+
+		$capability        = new Capability();
+		$register_response = $capability->register();
+	}
+
+	/**
+	 * Test administrator capabilities for Capability::MANAGE_REDIRECTS_CAPABILITY
+	 *
+	 * @return void
+	 */
+	public function test_new_admin_capability() {
+
+		$capability        = new Capability();
+		$register_response = $capability->register();
+
+		$user_id = $this->factory->user->create( array( 'role' => 'administrator' ) );
+		$user    = wp_set_current_user( $user_id );
+
+		$this->assertTrue( $user->has_cap( Capability::MANAGE_REDIRECTS_CAPABILITY ) );
```

Comment:
Let's wrap these assertions into custom assertion methods at the end of this test class. Something like:

```php
protected function assertUserHasRedirectCapability( $user ) {
	return $this->assertTrue( $user->has_cap( Capability::MANAGE_REDIRECTS_CAPABILITY ) );
}

protected function assertUserNotHasRedirectCapability( $user ) {
	return $this->assertFalse( $user->has_cap( Capability::MANAGE_REDIRECTS_CAPABILITY ) );
}

protected function assertRoleHasRedirectsCapability( $role ) {
	$user_id = $this->factory->user->create( array( 'role' => $role ) );
	$user    = wp_set_current_user( $user_id );

	return $this->assertUserHasRedirectCapability( $user );
}

protected function assertRoleNotHasRedirectsCapability( $role ) {
	$user_id = $this->factory->user->create( array( 'role' => $role ) );
	$user    = wp_set_current_user( $user_id );

	return $this->assertUserNotHasRedirectCapability( $user );
}
```

Then in the test methods, you can do:

```php
public function test_new_admin_capability() {
	$this->assertRoleNotHasRedirectsCapability( 'administrator' );

	$capability        = new Capability();
	$register_response = $capability->register();

	$this->assertRoleHasRedirectsCapability( 'administrator' );	
}
```

This could then be extracted to use a dataProvider since then the only variable is the role - wrap the final assertion in a condition that either checks the role has, or role has not got he cap, based on the expectation in the dataProvider.

Alternatively, just bundle the different role checks together:

```php
public function test_new_admin_capability() {
	$this->assertRoleNotHasRedirectsCapability( 'administrator' );
	$this->assertRoleNotHasRedirectsCapability( 'editor' );
	$this->assertRoleNotHasRedirectsCapability( 'subscriber' );

	$capability        = new Capability();
	$register_response = $capability->register();

	$this->assertRoleHasRedirectsCapability( 'administrator' );
	$this->assertRoleHasRedirectsCapability( 'editor' );
	$this->assertRoleNotHasRedirectsCapability( 'subscriber' ); // Should be no change.
}
```

Reply:
New methods added and the necessary code refactored